src/gf.c

```diff
@@ -914,8 +914,15 @@ DLLEXPORT jl_function_t *jl_instantiate_staged(jl_methlist_t *m, jl_tupletype_t
         }
     }
     func = with_appended_env(m->func, env);
-    jl_cellset(ex->args, 1, jl_apply(func, jl_svec_data(tt->parameters), jl_nparams(tt)));
+    jl_expr_t *body = jl_exprn(jl_symbol("block"), 2);
+    jl_cellset(ex->args, 1, body);
+    jl_expr_t *linenode = jl_exprn(line_sym, 2);
+    jl_cellset(body->args, 0, linenode);
+    jl_cellset(linenode->args, 0, jl_box_long(m->func->linfo->line));
+    jl_cellset(linenode->args, 1, m->func->linfo->file);
+    jl_cellset(body->args, 1, jl_apply(func, jl_svec_data(tt->parameters), jl_nparams(tt)));
     func = (jl_function_t*)jl_toplevel_eval_in(m->func->linfo->module, (jl_value_t*)ex);
+    func->linfo->name = m->func->linfo->name;
```

Comment:
yep probably safer. I think in this case linfo can never be NULL since we are evaling a lambda, but since being not null is not part of the "contract" of jl_function_t better be defensive.